R/check_spatvector.R

```diff
@@ -0,0 +1,145 @@
+#' Default is TRUE.
+#'
+#' @return The input SpatVector, possibly reprojected to EPSG:4326.
+#'
+#' @details
+#' The function checks:
+#' - If input is a SpatVector polygon
+#' - If coordinate system is properly defined
+#' - If polygons are valid and non-empty
+#' - If the area overlaps with known country boundaries
+#' - Optionally checks if area is within reasonable bounds
+#' (larger than 10,000 ha but smaller than approximately Brazil)
+#'
+#' @import terra
+#'
+#' @references
+#' Zillah Calle (2023)
+#' Jonas van Duijvenbode (2023)
+#'
+#' @examples
+#' \dontrun{
+#' shape <- vect("path/to/shapefile.shp")
+#' validated_shape <- check_spatvector(shape)
+#' }
+#'
+#' @export
+check_spatvector <- function(shape, check_size = TRUE) {
+  x <- check_basic_properties(shape)
+  x <- check_coordinate_system(shape)
+  x <- check_country_overlap(shape)
+  if (check_size) {
+    check_area_bounds(shape)
+  }
+  return(x)
+}
+
+#' Check basic SpatVector properties
+#' @noRd
+check_basic_properties <- function(x) {
+  if (!inherits(x, "SpatVector")) {
+    stop(paste0("Input must be a SpatVector object, received ", class(x)))
+  }
+  if (terra::geomtype(x) != "polygons") {
+    stop(paste("Input must be a polygon SpatVector, received", terra::geomtype(x)))
+  }
+  if (terra::nrow(x) == 0) {
+    stop("Input SpatVector contains no polygons")
+  }
+  return(x)
+}
+
+#' Check and fix coordinate system
+#' @noRd
+check_coordinate_system <- function(x) {
+  if (is.na(terra::crs(x))) {
+    x <- handle_missing_crs(x)
+  } else if (!is_wgs84(terra::crs(x))) {
+    ff_cat("Coordinate system is not EPSG:4326. Converting to WGS84.", color = "yellow")
+    x <- terra::project(x, "EPSG:4326")
+  }
+  return(x)
+}
+
+#' Handle missing CRS
+#' @noRd
+handle_missing_crs <- function(x) {
+  if (is_within_degree_range(x)) {
+    ff_cat("No coordinate system defined but coordinates appear to be in decimal degrees.
+            Assuming WGS84.", color = "yellow")
+    terra::crs(x) <- "EPSG:4326"
+    return(x)
+  }
+  stop("No coordinate system defined and coordinates are not in decimal degrees range")
+}
+
+#' Check if coordinates are within degree range
+#' @noRd
+is_within_degree_range <- function(x) {
+  terra::ext(x)$xmin >= -180 && terra::ext(x)$xmax <= 180
+}
+
+#' Check if CRS is WGS84
+#' @noRd
+is_wgs84 <- function(crs) {
+  identical(crs, "EPSG:4326") || length(grep("WGS 84", crs)) > 0
+}
+
+#' Check overlap with country boundaries
+#' @noRd
+check_country_overlap <- function(x) {
+  countries <- terra::aggregate(terra::vect(get(data("countries", envir = environment()))))
+  intersection <- terra::intersect(x, countries)
+
+  if (terra::nrow(intersection) == 0) {
+    stop("Input shape does not overlap with any known country boundaries")
+  }
+
+  check_overlap_percentage(x, intersection)
+  return(x)
+}
+
+#' Check overlap percentage
+#' @noRd
+check_overlap_percentage <- function(x, intersection) {
```

Comment:
You might want this function to return the overlap percentage in case you ever want to use that in some further analysis down the line. But I leave that up to you.

Reply:
did so, fixed, thanks!